### ISSUE TITLE
[Enhacement] StageManager alternative to be able to run autostage on non-active vessels.

### ISSUE THREAD

**Issue by jrodrigv** [MuMech/MechJeb2#1123]
Hi,

I think the title is enough descriptive but basically I think it would be very nice to be able to ascent vessels remotely even without focus on them.

Currently the StageManager is a limitation to achieve this because is bound to the active vessel and also couple with the UI.

**Comment by lamont-granquist** [comment]
I don't think you want to "write a custom StageManager" you just want to use the APIs that touch the vessel the MechJebCore is attached to.  I just don't know what APIs KSP offers to trigger staging on a particular Vessel object.  If you want to hunt around in the kOS codebase they might have hit and fixed this problem there.

**Comment by lamont-granquist** [comment]
A skim of kOS's docs suggest that kOS can only stage when the cpu vessel is the active vessel.

Also turned up this forum thread from years past:

https://forum.kerbalspaceprogram.com/index.php?/topic/64728-remote-staging/

That algorithm seems slightly dubious.

I'm not sure if that is what you meant by "write a custom StageManager".

**Comment by lamont-granquist** [comment]
kRPC also just uses StageManager.ActivateNextStage() on the active vessel.

this may be "easier" to implement as a feature to have MJ automatically switch to the active vessel for staging.

**Comment by lamont-granquist** [comment]
Although apparently the way to roll your own is:

- set vessel.updateResourcesOnEvent = false 
- loop through all the parts in the stage and fire OnStage on all of those parts
- set vessel.updateResourcesOnEvent = true
- call UpdateResourceSetsIfDirty
- then decrement the vessel's current stage

**Comment by sarbian** [comment]
I'll with the "power that be" if they are willing to move the staging code into Vessel. It would make sense to decouple it from the UI anyaway...

**Comment by sarbian** [comment]
Also, it is easy to duplicate the stock functionality in MJ (and I have done it many times) but the real problem is making sure that the code stays to to date with stock...

**Comment by lamont-granquist** [comment]
@sarbian yeah, that gives me pause as well...  best solution is if they'd just hang an interface off of Vessel in KSP itself...

**Comment by jrodrigv** [comment]
Yes I agree. I think KSP was not designed for remote staging although it support flybywire ? Lol. It is not very coherent.

BTW In BDArmory I implemented autostaging capabilities for misiles but based on action groups (each Custom Action group is one stage)  instead of StageManager.

**Comment by lamont-granquist** [comment]
did the really really trivially easy bit.

There's also `StageManager.LastStage` and `StageManager.StageCount` and `StageManager.ActivateStage(num)` plus still some `StageManager.CurrentStage` references in FuelFlowSimulation.

**Comment by jrodrigv** [comment]
I have been implementing the method of switching to the vessel that wants to stage and then switching back to the original active vessel.

It seems to work fine :) I still doing testing.

I will upload a demo video soon.

**Comment by jrodrigv** [comment]
@sarbian @lamont-granquist 
After few weeks of testing and fixing things. I don't only managed to make the autostage to work on remote vessels but also to enable the landing maneuver to be executed on different non-active vessels.
I have recorded a quite epic video as demo.
https://youtu.be/9jTMiJbdX7o

I have added a couple of new actions to Mechjeb. One for Landing somewhere and one to Land at target
I will open a Pull Request with all this stuff

**Comment by TheKurgan** [comment]
I really hope this gets integrated into MJ.
I have tested it as well, and it is an excellent addition.

**Comment by TheKurgan** [comment]
Sorry.... this was a year ago...
